grails-app/services/org/pih/warehouse/data/ProductSupplierDataService.groovy

```diff
@@ -141,7 +142,39 @@ class ProductSupplierDataService {
     }
 
     def getOrCreateNew(Map params) {
-        def productSupplier = params.productSupplier ? ProductSupplier.get(params.productSupplier) : null
+        def productSupplier
+        def data
+        if (params.productSupplier) {
+            productSupplier = params.productSupplier ? ProductSupplier.get(params.productSupplier) : null
+        } else {
+            String query = """
+                select 
```

Comment:
On the standup this morning I realized we could convert this query into a view that hides the complexity of what is happening. Essentially we'd return all of the same columns as the table, but we'd clean certain columns. We could honestly do this for a bunch of tables and maybe even use HQL/SQL to query the view and hydrate objects. I think you could do something like this https://blog.jdriven.com/2014/03/grails-goodness-using-hibernate-native-sql-queries/

```
def sessionFactory

def getOrCreateNew(Map params) { 
    sessionFactory.currentSession.createSQLQuery("select <columns> from product_supplier_view where <filters>") 
```

NOTE: I've started using a pattern to avoid the nasty branching logic. See the example below to see where you can basically ignore a filter by using `IFNULL(<user-param>, column_name)`

```
String forceRefreshStatement = String.format(
                            "delete from product_availability " +
                                    "where location_id = '${location.id}' " +
                                    "and product_id = IFNULL(%s, product_id);", productId)
                    stmt.addBatch(forceRefreshStatement)
```
This would be better as a prepared statement like you have below so see if you can figure that out.

Reply:
Here's some pseducode for a full solution
```
CREATE OR REPLACE VIEW product_supplier_clean_thingy AS (
    SELECT 
        id,
        code,
        name,
        supplier_id,
        REGEXP_REPLACE(supplier_code, '[ .,-]+', ''),
        ...
    FROM product_supplier
)

def getOrCreateNew(Map params) { 

    ProductSupplier productSupplier = getProductSupplier(params)
    if (!productSupplier) {
        productSupplier = createProductSupplierWihtoutPackage(params)
    }
    return productSupplier
}

def getProductSupplier(Map params) {

    final session = sessionFactory.currentSession

    // FIXME Eventually we'd like the regular expression to be configurable, 
    // but I'm not sure how that would work in the view above so we can 
    // ignore that for now
    String supplierCode = params.supplierCode ? params.supplierCode.replaceAll('[ .,-]+','') : null
    ...

    // Query string with :startId as parameter placeholder.
    final String query = """
        select id, ... 
        from product_supplier_clean_thingy  
        where product_id = :productId
        AND supplier_id = IFNULL(:supplierId, supplierId)
        AND supplier_code = IFNULL(:supplierCode, supplierCode)
        ...        
    """

    // Create native SQL query.
    final sqlQuery = session.createSQLQuery(query)

    // Use Groovy with() method to invoke multiple methods
    // on the sqlQuery object. 
    // FIXME Not sure if this works with our version but I've seen similar solutions with Grails 1.3.9
    final results = sqlQuery.with {
        // Set domain class as entity.
        // Properties in domain class id, name, level will
        // be automatically filled.
        addEntity(ProductSupplier)

        // Set value for parameter startId.
        setString('productId', productId)
        setString('supplierCode', supplierCode)

        // Get all results.
        // FIXME Not sure if you can use get() or max/limit here. If not we'll just check 
        // the results to see if there's more than one and return the first result
        list()
    }

    results
}
```

Reply:
I've tried creating native sql query but it didn't work. I had weird errors that "version" column does not exist and also some other strange. I've updated PR with other changes that you have suggested.